### DISCUSSION TITLE
Welcome to robotics-toolbox-python Discussions!

### DISCUSSION THREAD

**Discussion by petercorke** [petercorke/robotics-toolbox-python#158] [General]
## 👋 Welcome!
  The Robotics Toolbox for Python is a new project, inspired by its MATLAB version, and we're still actively working on it.
  We’re using Discussions as a place to connect with users and potential contributors. We hope that you:
  * Ask general questions about the project and where it's going.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

This is **not the place** for bug reports.  For bugs please create an [Issue](https://github.com/petercorke/robotics-toolbox-python/issues/new/choose)

**Comment by ffaria431** [comment]
Hi Peter,

I have found two errors to run the code in Kinematics.ipynb file:
Error 1. "Could not connect to the Swift simulator" during running the code below 
env = rtb.backends.Swift()  # instantiate 3D browser-based visualizer
env.launch()                # activate it
env.add(robot)              # add robot to the 3D scene
for qk in qt.q:             # for each joint configuration on trajectory
      robot.q = qk          # update the robot state
      env.step()            # update visualization

Error 2. The line below do not run,
p560.plot(p560.qn);

Can you please provide your valuable advise regarding these problems?
Thanks in advance!

**Comment by petercorke** [comment]
Sent from my iPad

> On 17 Mar 2021, at 8:40 am, ffaria431 ***@***.***> wrote:
> 
> ﻿
> Hi Peter,
> 
> I have found two errors to run the code in Kinematics.ipynb file:
> Error 1. "Could not connect to the Swift simulator" during running the code below
> env = rtb.backends.Swift() # instantiate 3D browser-based visualizer
> env.launch() # activate it
> env.add(robot) # add robot to the 3D scene
> for qk in qt.q: # for each joint configuration on trajectory
> robot.q = qk # update the robot state
> env.step() # update visualization
> 

This is due to recent backend changes, easiest way to use Swift is the 

backend=‘swift’

Option to plot.  
> Error 2. The line below do not run,
> p560.plot(p560.qn);
> 

What was the error?

**Reply by ffaria431** [reply]
Error was "IPython is not defined".

Moreover, during running qt.q , this error is occurred :  "'Trajectory' object has no attribute 'q'" . Thanks.

**Comment by petercorke** [comment]
Pls give me a complete example that shows the problem 

Peter.

**Reply by ffaria431** [reply]
![Screen Shot 2021-03-17 at 1 04 16 PM](https://user-images.githubusercontent.com/57658942/111507739-5b474500-8721-11eb-88cf-5225185adcf2.png)
I have attached the screenshot here. I am trying to run kinematics.ipynb file of your code. line 5 and 6 is my problem. Line 6 don't show output.

**Comment by ffaria431** [comment]
I have another query. I am working with C-Arm kinematics. I need to create DH table for 10 DOF for C-Arm. Do you have any materials or code to create DH parameter for own piece of Robot? Thanks a million. I really appreciate your worthy time and assistance.

**Comment by Matt-Collins-52** [comment]
When calculating the inverse kinematics, the program returns a list of numbers representing the joint angles. How would I call these angles individually so they can be converted into something like motor movement? I cannot find what their variable names are.

**Reply by Matt-Collins-52** [reply]
I guess I put this out a little too soon and found out shortly after. For anyone wondering the same thing, the individual joint angles from the inverse kinematics can be called by using sol.q[n] where n is the joint number.

**Comment by petercorke** [comment]
Correct!  The object contains other useful status information as well.

The kinematic notebook and the top level README explain this

**Comment by Matt-Collins-52** [comment]
When creating a new DH model, I can enter alpha, d, and a, but I cannot enter the theta column from my DH table. Is there a way to do this or a different way around this?

**Reply by Adineh94** [reply]
Hi MCollins,
Once you want to create a rigid body tree, "alpha, d and,  a" are required, but "theta" is the input parameters.

**Comment by Adineh94** [comment]
**Question: How to Import my robot.URDF to python in such a way to get all properties (M, C, J, ...) same as MATLAB Robotics Toolbox.**

Explanation:
Because of some reason, I have to move to python.
It was very easy and straightforward to import our serial robot(URDF format) in Matlab, for example:

clc
clear
robotCadName = 'mySerialRobot.urdf';
[folderName,~] = split(robotCadName,'.');
meshPath = fullfile(pwd, 'CADModel', 'URDF', folderName1{1});
robot = importrobot(robotCadName,'MeshPath', meshPath);
clearCollision(robot.Base)
clearVisual(robot.Base)

I think It can be done manually by defining DH and Mechanical properties to create DHRobot. However, a simple way would be much appreciated.

**Comment by Matt-Collins-52** [comment]
I'm having some trouble with plotting. I have tried different ways to plot several robots, and I usually end up with this error:

    AttributeError: 'Trajectory' object has no attribute 'q'

Python says the error comes from these lines.

    qt = rtb.jtraj(robot.qz, q_pickup, 50)
    robot.plot(qt.q, movie='panda1.gif')

This also shows a blank plot with no robot in it. Does anyone know how to fix this?

**Comment by petercorke** [comment]
Please, you need to be more precise in what you write if you want help.

> On 17 Mar 2021, at 8:40 am, ffaria431 ***@***.***> wrote:
> 
> 
> Hi Peter,
> 
> I have found two errors to run the code in Kinematics.ipynb file:
> Error 1. "Could not connect to the Swift simulator" during running the code below
> env = rtb.backends.Swift() # instantiate 3D browser-based visualizer
> env.launch() # activate it
> env.add(robot) # add robot to the 3D scene
> for qk in qt.q: # for each joint configuration on trajectory
> robot.q = qk # update the robot state
> env.step() # update visualization
> 
> 
there is no code like this in the .ipynb file
> Error 2. The line below do not run,
> p560.plot(p560.qn);
> 
> 
as in the fourth cell in the notebook?  Send error messages.

**Comment by petercorke** [comment]
How are you creating the trajectory object?

**Reply by Matt-Collins-52** [reply]
I just followed the readme and got the attribute error stated before. Sometimes I can also change some things and get no errors, but still get a blank graph either way.

**Reply by petercorke** [reply]
Can you send a complete code snippet.

**Reply by petercorke** [reply]
(rtb) >>> tg=jtraj(puma.qr, puma.qz, 20)                                        

(rtb) >>> tg.q                                                                  
Out[3]: 
array([[       0,    1.571,   -1.571,        0,        0,        0],
       [       0,    1.569,   -1.569,        0,        0,        0],
       [       0,    1.555,   -1.555,        0,        0,        0],
       [       0,    1.523,   -1.523,        0,        0,        0],
       [       0,    1.467,   -1.467,        0,        0,        0],
       [       0,    1.386,   -1.386,        0,        0,        0],
       [       0,    1.281,   -1.281,        0,        0,        0],
       [       0,    1.155,   -1.155,        0,        0,        0],
       [       0,    1.014,   -1.014,        0,        0,        0],
       [       0,   0.8628,  -0.8628,        0,        0,        0],
       [       0,    0.708,   -0.708,        0,        0,        0],
       [       0,   0.5567,  -0.5567,        0,        0,        0],
       [       0,   0.4154,  -0.4154,        0,        0,        0],
       [       0,   0.2899,  -0.2899,        0,        0,        0],
       [       0,   0.1852,  -0.1852,        0,        0,        0],
       [       0,   0.1042,  -0.1042,        0,        0,        0],
       [       0,  0.04811, -0.04811,        0,        0,        0],
       [       0,  0.01555, -0.01555,        0,        0,        0],
       [       0, 0.002113, -0.002113,        0,        0,        0],
       [       0,        0,        0,        0,        0,        0]])

**Reply by Matt-Collins-52** [reply]
Here is what I have from the readme

  import roboticstoolbox as rtb
  from spatialmath import SE3
  
  robot = rtb.models.DH.Panda()
  
  T = robot.fkine(robot.qz)  # forward kinematics
  
  T = SE3(0.7, 0.2, 0.1) * SE3.OA([0, 1, 0], [0, 0, -1])
  sol = robot.ikine_LM(T)         # solve IK
  
  q_pickup = sol.q
  print(robot.fkine(q_pickup))    # FK shows that desired end-effector pose was achieved
  
  qt = rtb.jtraj(robot.qz, q_pickup, 50)
  robot.plot(qt.q, movie='panda1.gif')

**Reply by petercorke** [reply]
Here's what would have saved a lot of time.  You send me the error traceback like below for the simplest case that will cause the problem.

```
>>> robot.plot(robot.qz)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "code/robotics-toolbox-python/roboticstoolbox/robot/Robot.py", line 1375, in plot
    env.add(
  File "code/robotics-toolbox-python/roboticstoolbox/backends/PyPlot/PyPlot.py", line 281, in add
    self.robots[len(self.robots) - 1].draw()
  File "code/robotics-toolbox-python/roboticstoolbox/backends/PyPlot/RobotPlot.py", line 79, in draw
    self.init()
  File "code/robotics-toolbox-python/roboticstoolbox/backends/PyPlot/RobotPlot.py", line 204, in init
    shadow, = self.ax.plot(
  File "opt/anaconda3/lib/python3.8/site-packages/mpl_toolkits/mplot3d/axes3d.py", line 1421, in plot
    zs = np.broadcast_to(zs, len(xs))
TypeError: object of type 'int' has no len()
```

The problem I'm seeing is in matplotlib from plot, nothing to do with trajectories.  I'll check this out over the weekend.

**Reply by Matt-Collins-52** [reply]
The error traceback I get from my code above is only one line, which is also above. It says the trajectory object has no attribute q. Everything is in the first comment I made.

I also changed qt.q to qt.y and get no errors, but there is still a blank plot showing up. I can also now print the trajectory so the problem is with plotting.

**Comment by petercorke** [comment]
robot = ERobot.URDF(urdf file name)

should create an ERobot instance from the file.  Inertial information, if present will be read in.  You can display the kinematic params
by

print(robot)

and the inertial params by

robot.dynamics()

peter

**Reply by Adineh94** [reply]
Hi Peter, Many Thanks!

**Error:**
I have tested, but I have received the following error:
AttributeError: type object 'ERobot' has no attribute 'URDF'

**My Code:**
from roboticstoolbox import ERobot
import os

pwd = os.getcwd()
pth = os.path.dirname(pwd)
urdf_loc = pth + "\\two_links_robot\\two_links_robot.urdf"

robot = ERobot.URDF(urdf_loc)

**Extra Explanation:**
As a simple example, I've created a two links robot by Solidwork and export it as URDF. That creates a URDF as a package (For example "mesh filename="package://two_links_robot/meshes/Base.STL" />" for Base) which directly can be loaded to MATLAB. However, I have found this should be modified for some URDF readers such as urdfpy. Both files have been attached.
+++++++++++++++++++++++++++++++ Code ++++++++++++
from urdfpy import URDF
import os
import numpy as np

pwd = os.getcwd()
pth = os.path.dirname(pwd)
urdf_loc = pth + "\\two_links_robot\\two_links_robot.urdf"
robot = URDF.load(urdf_loc)
robot.animate(cfg_trajectory={
    'q0': [-np.pi / 4, np.pi / 4],
    'q1': [0.0, -np.pi / 2.0],
})

[two_links_robot - Modified.zip](https://github.com/petercorke/robotics-toolbox-python/files/6402967/two_links_robot.-.Modified.zip)
[two_links_robot.zip](https://github.com/petercorke/robotics-toolbox-python/files/6402968/two_links_robot.zip)

**Comment by petercorke** [comment]
What version are you running, pypi or GitHub?

**Reply by Matt-Collins-52** [reply]
I'm not sure. Is there any way to check this?

**Comment by petercorke** [comment]
Did you install it with Pip or git clone?

**Reply by Matt-Collins-52** [reply]
I believe git clone, but I tried both ways and had some trouble and eventually got it.

**Comment by petercorke** [comment]
The version on GH has the ability to read URDF


ERobot.URDF(filename)


You don’t need urdfpy.

Peter

**Reply by Adineh94** [reply]
Hi Peter, Thanks again!
I have found GH version can not recognize "quadprog" and "qpsolver" on Windows 10 and I've received the attached errors. However, I was able to completely install it on ubuntu. 
Now I am able to create my robot object from the URDF file. However, I have to copy the URDF package on rtb-data!
**New Error:
Once I want to calculate "acc", "tau" and M, C, G matrix, I have received this error:**

s[j] = link.v.s
AttributeError: 'NoneType' object has no attribute 's'

Which seems related to line 2105 of the ERobot module.
#################error: installing on windows ####################
 ERROR: Command errored out with exit status 1:
   command: 'C:\Users\moham\miniconda3\envs\rtb\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\moham\\AppData\\Local\\Temp
\\pip-install-3f548fr0\\quadprog_e756ac553e55447faa445d89b72774ea\\setup.py'"'"'; __file__='"'"'C:\\Users\\moham\\AppData\\Local\\Temp\\pip-install-3f548fr0\\
quadprog_e756ac553e55447faa445d89b72774ea\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"
');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\moham\AppData\Local\Temp\pip-wheel-i7s52nqp'
       cwd: C:\Users\moham\AppData\Local\Temp\pip-install-3f548fr0\quadprog_e756ac553e55447faa445d89b72774ea\
  Complete output (6 lines):
  running bdist_wheel
  running build
  running build_ext
  skipping 'quadprog\quadprog.cpp' Cython extension (up-to-date)
  building 'quadprog' extension
  error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/

  ----------------------------------------
  ERROR: Failed building wheel for quadprog





  ERROR: Command errored out with exit status 1:
     command: 'C:\Users\moham\miniconda3\envs\rtb\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\moham\\AppData\\Local\\Te
mp\\pip-install-3f548fr0\\quadprog_e756ac553e55447faa445d89b72774ea\\setup.py'"'"'; __file__='"'"'C:\\Users\\moham\\AppData\\Local\\Temp\\pip-install-3f548fr0
\\quadprog_e756ac553e55447faa445d89b72774ea\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"
'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\moham\AppData\Local\Temp\pip-record-t84keeva\install-record.txt' --si
ngle-version-externally-managed --compile --install-headers 'C:\Users\moham\miniconda3\envs\rtb\Include\quadprog'
         cwd: C:\Users\moham\AppData\Local\Temp\pip-install-3f548fr0\quadprog_e756ac553e55447faa445d89b72774ea\
    Complete output (6 lines):
    running install
    running build
    running build_ext
    skipping 'quadprog\quadprog.cpp' Cython extension (up-to-date)
    building 'quadprog' extension
    error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tool
s/
    ----------------------------------------
ERROR: Command errored out with exit status 1: 'C:\Users\moham\miniconda3\envs\rtb\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\
\Users\\moham\\AppData\\Local\\Temp\\pip-install-3f548fr0\\quadprog_e756ac553e55447faa445d89b72774ea\\setup.py'"'"'; __file__='"'"'C:\\Users\\moham\\AppData\\
Local\\Temp\\pip-install-3f548fr0\\quadprog_e756ac553e55447faa445d89b72774ea\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().
replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\moham\AppData\Local\Temp\pip-record-
t84keeva\install-record.txt' --single-version-externally-managed --compile --install-headers 'C:\Users\moham\miniconda3\envs\rtb\Include\quadprog' Check the l
ogs for full command output.

**Comment by petercorke** [comment]
I'd prefer that discussions like this happen in Issues not here in discussion.  Issues is much better suited for that.  I will delete all these posts in a few ays.

**Comment by kraudust** [comment]
@petercorke Is there a way to do self collision checking? I see the iscollided function, but it seems this is for checking if any of the arm links have collided with an external shape as you have to pass in the joint angles and a shape. But I'm wondering if there is a way to check if any of the arm links are colliding with each other in a certain joint angle configuration. I've checked that my links have collision geometry, and even tried passing in a link geometry to the iscollided function, but it always returns true in this case since this function cycles through the links and checks if they have collided with the shape (in this case, a link will have collided with itself).

**Reply by Ruthrash** [reply]
Hi @kraudust. self-collision check has been imlemented for Panda in DH models. Make sure to have install prtb with collision and try the following code snippet. 

```
import roboticstoolbox as rtb 

panda = rtb.models.DH.Panda() 

panda.collided(panda.q, panda) 
```

**Comment by TheProbPro** [comment]
@petercorke im currently trying to setup a simulation of a UR3 robot arm through the SWIFT backend. Im currently running into an issue where the compiler tells me that there is no running asyncio eventloops. Im running some code i know worked back in august of 2024, so i tried reverting all relevant packages to then, but i still get the same issue. Do you have any suggestions on how to fix this?

Traceback (most recent call last):
  File "C:\AppData\Local\Programs\Python\Python313\Lib\threading.py", line 1041, in _bootstrap_inner
    self.run()
    ~~~~~~~~^^
  File "C:\AppData\Local\Programs\Python\Python313\Lib\threading.py", line 992, in run
    self._target(*self._args, **self._kwargs)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Documents\.venv\Lib\site-packages\swift\SwiftRoute.py", line 296, in __init__
    start_server = websockets.serve(self.serve, "localhost", port)
  File "C:\Documents\.venv\Lib\site-packages\websockets\asyncio\server.py", line 743, in __init__
    self.server = Server(
                  ~~~~~~^
        handler,
        ^^^^^^^^
    ...<4 lines>...
        logger=logger,
        ^^^^^^^^^^^^^^
    )
    ^
  File "C:\Documents\.venv\Lib\site-packages\websockets\asyncio\server.py", line 284, in __init__
    self.loop = asyncio.get_running_loop()
                ~~~~~~~~~~~~~~~~~~~~~~~~^^
RuntimeError: no running event loop
